allennlp/training/callbacks/callbacks.py

```diff
@@ -0,0 +1,555 @@
+                self.tensorboard.add_train_scalar("current_batch_size", cur_batch)
+                self.tensorboard.add_train_scalar("mean_batch_size", average)
+
+        if self.tensorboard.should_log_histograms_this_batch():
+            for name, param in trainer.model.named_parameters():
+                self.param_updates[name].sub_(param.detach().cpu())
+                update_norm = torch.norm(self.param_updates[name].view(-1, ))
+                param_norm = torch.norm(param.view(-1, )).cpu()
+                self.tensorboard.add_train_scalar("gradient_update/" + name,
+                                                  update_norm / (param_norm + 1e-7))
+            self.param_updates.clear()
+            self.tensorboard.log_histograms(trainer.model, self.histogram_parameters)
+
+    @handle_event(Events.EPOCH_END)
+    def epoch_end_logging(self, trainer):
+        self.tensorboard.log_metrics(trainer.train_metrics,
+                                     val_metrics=trainer.val_metrics,
+                                     log_to_console=True,
+                                     epoch=trainer.epoch_number + 1)
+
+    @classmethod
+    def from_params(cls, serialization_dir: str, params: Params) -> 'LogToTensorboard':  # type: ignore
+        log_batch_size_period = params.pop_int("log_batch_size_period", None)
+        tensorboard = TensorboardWriter.from_params(params=params,
+                                                    serialization_dir=serialization_dir,
+                                                    get_batch_num_total=lambda: None)
+        return LogToTensorboard(tensorboard, log_batch_size_period)
+
+
+@Callback.register("learning_rate_scheduler")
+class LrsCallback(Callback):
+    """
+    Callback that handles the learning rate scheduler.
+    """
+    def __init__(self, learning_rate_scheduler: LearningRateScheduler) -> None:
+        self.learning_rate_scheduler = learning_rate_scheduler
+
+    @handle_event(Events.AFTER_BACKWARD)
+    def step_batch(self, trainer):
+        self.learning_rate_scheduler.step_batch(trainer.batch_num_total)
+
+    @handle_event(Events.EPOCH_END)
+    def step(self, trainer):
+        self.learning_rate_scheduler.step(trainer.latest_val_metric, trainer.epoch_number)
+
+    def get_training_state(self) -> dict:
+        """
+        We need to persist the learning_rate_scheduler state as training state.
+        """
+        return {"learning_rate_scheduler": self.learning_rate_scheduler.state_dict()}
+
+    def restore_training_state(self, training_state: dict) -> None:
+        state_dict = training_state.pop("learning_rate_scheduler", None)
+
+        if state_dict:
+            self.learning_rate_scheduler.load_state_dict(state_dict)
+
+
+@Callback.register("momentum_scheduler")
+class MomentumSchedulerCallback(Callback):
+    """
+    Callback that runs a Momentum Scheduler.
+    TODO: this looks almost exactly like the LrsCallback, could we share code?
```

Comment:
Probably not, but either way, the TODO should be in a comment, not in the docstring (which goes into our docs).